walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,217 @@
 
 package com.google.sps;
 
+import java.lang.Math;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+    // If there are no mandatory attendees, the whole day is free for optional
+    if (request.getAttendees().isEmpty()) {
+      meetingTimes.add(TimeRange.WHOLE_DAY);
+    }
+    // If the meeting duration is longer than a day, there is no slot available
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    Map<String, ArrayList<TimeRange>> schedules = new HashMap<String, ArrayList<TimeRange>>();
+    schedules = initSchedules(request, schedules);
+
+    // Update each attendee's schedule based on events
+    for (Event event : events) {
+      for (String attendee : event.getAttendees()) {
+        if (schedules.containsKey(attendee)) {
+          for (ListIterator scheduleIterator = schedules.get(attendee).listIterator();
+               scheduleIterator.hasNext();) {
+            schedules = scheduleEvent(event, attendee, scheduleIterator, schedules);
+          }
+        }
+      }
+    }
+
+    // Takes into consideration all attendees to find free slots
+    int firstFlag = 1;
+    ArrayList<TimeRange> meetingTimesToAdd = new ArrayList<TimeRange>();
+    for (String attendee : request.getAttendees()) {
+      meetingTimesToAdd = determineAvailableTime(
+          request, firstFlag, meetingTimesToAdd, schedules, attendee, meetingTimes);
+      for (TimeRange adding : meetingTimesToAdd) {
+        meetingTimes.add(adding);
+      }
+      meetingTimesToAdd.clear();
+      firstFlag = 0;
+    }
+
+    // Takes into consideration all optional attenddes to find free slots
+    ArrayList<TimeRange> optionalMeetingTimesToAdd = new ArrayList<TimeRange>();
+    Map<String, ArrayList<TimeRange>> optionalFrees = new HashMap<String, ArrayList<TimeRange>>();
+    for (String optionalAttendee : request.getOptionalAttendees()) {
+      firstFlag = 0;
+      ArrayList<TimeRange> optionalMeetingTimes = (ArrayList<TimeRange>) meetingTimes.clone();
+      optionalMeetingTimesToAdd = determineAvailableTime(request, firstFlag,
+          optionalMeetingTimesToAdd, schedules, optionalAttendee, optionalMeetingTimes);
```

Comment:
I'm curious why you need to return optionalMeetingTimesToAdd if you pass it as a parameter. The one in the parameter should be changed when the call returns.